src/types.jl

```diff
@@ -38,24 +38,61 @@ SCSMatrix(m::ManagedSCSMatrix) =
-    verbose::Int # boolean, write out progress: 1
-    warm_start::Int # boolean, warm start (put initial guess in Sol struct): 0
-    acceleration_lookback::Int # acceleration memory parameter: 20
+    normalize::Int # boolean, heuristic data rescaling
+    scale::Cdouble # if normalized, rescales by this factor
+    rho_x::Cdouble # x equality constraint scaling
+    max_iters::Int # maximum iterations to take
+    eps::Cdouble # convergence tolerance
+    alpha::Cdouble # relaxation parameter
+    cg_rate::Cdouble # for indirect, tolerance goes down like (1/iter)^cg_rate
+    verbose::Int # boolean, write out progress
+    warm_start::Int # boolean, warm start (put initial guess in Sol struct)
+    acceleration_lookback::Int # acceleration memory parameter
+
+    SCSSettings() = new()
+    SCSSettings(normalize, scale, rho_x, max_iters, eps, alpha, cg_rate, verbose, warm_start, acceleration_lookback) = new(normalize, scale, rho_x, max_iters, eps, alpha, cg_rate, verbose, warm_start, acceleration_lookback)
 end
 
-function SCSSettings(;normalize=1::Int, scale=convert(Cdouble, 1.0)::Cdouble, rho_x=convert(Cdouble,1e-3)::Cdouble,
-                        max_iters=5000::Int, eps=convert(Cdouble, 1e-5)::Cdouble, alpha=convert(Cdouble, 1.5)::Cdouble,
-                        cg_rate=convert(Cdouble,2)::Cdouble, verbose=1::Int, warm_start=0::Int, acceleration_lookback=20::Int)
-    return SCSSettings(normalize, scale, rho_x, max_iters, eps, alpha, cg_rate, verbose, warm_start, acceleration_lookback)
+struct Direct end
+struct Indirect end
+
+function _SCS_user_settings(default_settings;
+        normalize=default_settings.normalize,
+        scale=default_settings.scale,
+        rho_x=default_settings.rho_x,
+        max_iters=default_settings.max_iters,
+        eps=default_settings.eps,
+        alpha=default_settings.alpha,
+        cg_rate=default_settings.cg_rate,
+        verbose=default_settings.verbose,
+        warm_start=default_settings.warm_start,
+        acceleration_lookback=default_settings.acceleration_lookback)
+    return SCSSettings(normalize, scale, rho_x, max_iters, eps, alpha, cg_rate, verbose,warm_start, acceleration_lookback)
 end
 
+function SCSSettings(linear_solver::Union{Type{Direct}, Type{Indirect}}; options...)
+
+    mmatrix = ManagedSCSMatrix(0,0,spzeros(1,1))
+    matrix = Ref(SCSMatrix(mmatrix))
+    default_settings = Ref(SCSSettings())
+    dummy_data = Ref(SCSData(0,0, Base.unsafe_convert(Ptr{SCSMatrix}, matrix),
+        pointer([0.0]), pointer([0.0]),
+        Base.unsafe_convert(Ptr{SCSSettings}, default_settings)))
+    SCS_set_default_settings(linear_solver, dummy_data)
+    try
+        settings = _SCS_user_settings(default_settings[]; options...)
+    catch err
+        if err isa MethodError
+            SCS_kwargs = fieldnames(default_settings[])
+            kwargs = [err.args[1][i*2-1] for i in 1:(length(err.args[1]) ÷ 2)]
```

Comment:
There are a number of ways that this code could stop working unintentionally, so please add tests. The structure of `err` could be changed across Julia versions or `_SCS_user_settings` could be changed.

Reply:
I can `@test_throws ArgumentError ...` against misspelled options, but that's it. 
`_SCS_user_settings` is an internal function, it won't change beyond our control (btw. why is `SCSSettings` exported? I had always assumed that user is not supposed to construct it?)

I'd use the standard `MethodError` which actually tells exactly which options were not recognized, but that clashes with the style guide. What are Your suggestions on handling this?

Reply:
> I can @test_throws ArgumentError ... against misspelled options, but that's it.

That would make good tests :)

> it won't change beyond our control

yes but when we change it we might forget to change `MOIWrapper.jl`.